doc/fpmsyncd/hld_fpmsyncd-NTT.md

```diff
@@ -0,0 +1,320 @@
+
+<table>
+  <tr><td>Event</td><td>Attributes</td><td>Description</td></tr>
+  <tr><td rowspan="3">RTM_NEWNEXTHOP</td><td>NHA_ID</td><td>NextHop Group ID</td></tr>
+  <tr><td>NHA_GATEWAY</td><td>gateway address</td></tr>
+  <tr><td>NHA_OIF</td><td>The interface ID</td></tr>
+  <tr><td rowspan="2">RTM_NEWNEXTHOP</td><td>NHA_ID</td><td>NextHop Group ID</td></tr>
+  <tr><td>NHA_GROUP</td><td>A list of nexthop groups IDs with its respective weights.</td></tr>
+</table>
+
+After sending the `RTM_NEWNEXTHOP` events, zebra sends the `RTM_NEWROUTE` to `fpmsyncd` with NextHop Group ID as shown in the table below:
+
+<table>
+  <tr><td>Event</td><td>Attributes</td><td>Description</td></tr>
+  <tr><td rowspan="2">RTM_NEWROUTE</td><td>RTA_DST</td><td>route prefix address</td></tr>
+  <tr><td>RTA_NH_ID</td><td>NextHop Group ID</td></tr>
+</table>
+
+#### Example of entries in APPL_DB
+
+For example. following route configuration will generate events show in the table below:
+
+```
+S>* 8.8.8.0/24 [1/0] via 10.0.1.5, Ethernet4, weight 1, 00:01:09 
+  * via 10.0.2.6, Ethernet8, weight 1, 00:01:09 
+S>* 9.9.9.0/24 [1/0] via 10.0.1.5, Ethernet4, weight 1, 00:00:04 
+  * via 10.0.2.6, Ethernet8, weight 1, 00:00:04
+```
+<table>
+  <tr><td>Seq</td><td>Event</td><td>Attributes</td><td>Value</td></tr>
+  <tr><td rowspan="3">1</td><td rowspan="3">RTM_NEWNEXTHOP</td><td>NHA_ID</td><td>116</td></tr>
+  <tr><td>NHA_GATEWAY</td><td>10.0.1.5</td></tr>
+  <tr><td>NHA_OIF</td><td>22</td></tr>
+  <tr><td rowspan="3">2</td><td rowspan="3">RTM_NEWNEXTHOP</td><td>NHA_ID</td><td>117</td></tr>
+  <tr><td>NHA_GATEWAY</td><td>10.0.2.6</td></tr>
+  <tr><td>NHA_OIF</td><td>23</td></tr>
+  <tr><td rowspan="2">3</td><td rowspan="2">RTM_NEWNEXTHOP</td><td>NHA_ID</td><td>118</td></tr>
+  <tr><td>NHA_GROUP</td><td>[{116,1},{117,1}]</td></tr>
+  <tr><td rowspan="2">4</td><td rowspan="2">RTM_NEWROUTE</td><td>RTA_DST</td><td>8.8.8.0/24</td></tr>
+  <tr><td>RTA_NH_ID</td><td>118</td></tr>
+  <tr><td rowspan="2">5</td><td rowspan="2">RTM_NEWROUTE</td><td>RTA_DST</td><td>9.9.9.0/24</td></tr>
+  <tr><td>RTA_NH_ID</td><td>118</td></tr>
+</table>
+
+A short description of `fpmsyncd` logic flow:
+
+- When receiving `RTM_NEWNEXTHOP` events on sequence 1, 2 and 3, `fpmsyncd` will save the information in an internal list to be used when necessary.
+- When `fpmsyncd` receive the `RTM_NEWROUTE` on sequence 4, the process will write the NextHop Group with ID 118 to the `NEXTHOP_GROUP_TABLE` using the information of gateway and interface from the NextHop Group events with IDs 116 and 117.
+- Then `fpmsyncd` will create a new route entry to `ROUTE_TABLE` with a `nexthop_group` field with value `ID118`.
+- When `fpmsyncd` receives the last `RTM_NEWROUTE` on sequence 5, the process will create a new route entry (but no NextHop Group entry) in `ROUTE_TABLE` with `nexthop_group` field with value `ID118`. (Note: This NextHop Group entry was created when the `fpmsyncd` received the event sequence 4.)
+
```

Comment:
Will Zebra/FRR re-create the NHG (delete old one + create new one) and update all routes with new ID on change of NHG (e.g. due to failed link to NHGm) ?

Reply:
Yes, if there are one or more route referring to the NHG with members before the change.
(In this case we will have an old NHG with n members referenced from some routes and a newly created NHG n-x members referenced from other routes)

I think it's No if there are no routes referencing to the NHG with members before the change.
But I'm not 100% confident so let me check again and update the result here.

Reply:
That would depend on NHG type (BGP vs IGP),  the type of triggered events and how the trigger events get handled in zebra.

Reply:
I have added explanation that NHG ID and members are managed by FRR in "Overview" section for clarification.